Week 3 / Day 6 / Test Automation (#72)
microsoft/win-student-devs: docs/30DaysOfPWA/dev-tools/06.md

# 3.6: Test automation tools

**Author: Nitya Narasimhan [@nitya](https://twitter.com/nitya)**

Welcome to Week 3, **Day 6** of #30DaysOfPWA. Check out our [Kickoff](../kickoff.md) post to learn about our roadmap and contributors. In today's post we'll talk about _Test Automation_ - what it is, why it matters, and developer tools to help us build reliable end-to-end testing workflows for our PWA.

![image of title and author.](_media/day-06.jpg)

### WHAT WE'LL COVER TODAY

| Section | Description |
| ------- | ----------- |
| **Test Automation** | What is it and why does it matter for modern web apps?|
|**Test Pyramid** | What does a testing strategy for modern web apps involve? |
|**Testing PWA** | What does a functional testing spec for PWA need? |
|**Hello, Playwright**| Reliable end-to-end test framework for modern web apps. |
|**Using Playwright** | Explore quickstarts, powerful tools, and PWA test scenarios |
|**Exercise**| Take Playwright for a spin with a sample PWA! |

---

## Test Automation

> Test Automation is the practice of _executing tests automatically_ to validate software product functionality, then _using testing insights_ to iteratively improve product quality. Test Automation is key to continuous delivery in modern DevOps culture, using build-deploy pipelines maintain software products in a continuous state of readiness for production release. 

Reviewer: ```suggestion
Test Automation is the practice of _executing tests automatically_ to validate software product functionality, then _using testing insights_ to iteratively improve product quality. Test Automation is key to continuous delivery in modern DevOps culture: build-deploy pipelines keep software products in a continuous state of readiness for production release. 
```

Why explore Test Automation for PWA? Two reasons: growing test complexity & better tooling capability.

#### Testing Complexity

Progressive Web Apps combine the reach of traditional web apps with the richer capability of platform-native apps. The first needs _cross-browser_ testing to ensure usable experiences across all targeted browsers. The second needs _progressive enhancement_ testing strategies to validate the enhanced experiences available to more capable devices.

> Increased frequency of testing (e.g., on every feature commit or bug fix) coupled with many testing granularities (unit, integration, end-to-end) and configurations (browsers, device platforms) makes _manual testing_ time-consuming and error-prone. This is where test automation excels, reducing time to detect issues, eliminating operator errors. and scaling effectively with cloud resources.

Reviewer: ```suggestion
Increased frequency of testing (e.g., on every feature commit or bug fix) coupled with many testing granularities (unit, integration, end-to-end) and configurations (browsers, device platforms) makes manual testing time-consuming and error-prone. This is where test automation excels, reducing time to detect issues, eliminating operator errors, and scaling effectively with cloud resources.
```

#### Tooling Capability

> Adopting automated testing strategies today is a no-brainer thanks to the rich ecosystem of tools and technologies available to web developers. The game changer is _web automation_ software that can execute user actions and other operations on a web browser without human intervention. Check out the [Microsoft Edge docs](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/test-and-automation/test-and-automation) for more on relevant testing tools and technologies for web development, including:

Reviewer: ```suggestion
Adopting automated testing strategies today is a no-brainer thanks to the rich ecosystem of tools and technologies available to web developers. Modern web automation software can execute user actions and other operations on a web browser without human intervention. Check out the [Microsoft Edge docs](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/test-and-automation/test-and-automation) for more on relevant testing tools and technologies for web development, including:
```

>  * _[Origin Trials](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/origin-trials)_ - trial experimental APIs (e.g., web capabilities) on live sites with real users.

Author: ```suggestion
```

>  * _[Origin Trials](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/origin-trials)_ - trial experimental APIs (e.g., web capabilities) on live sites with real users.

Author: I'm actually good with removing it. It was on the list specified in https://docs.microsoft.com/en-us/microsoft-edge/test-and-automation/test-and-automation (Tests and Automation page on Edge Docs) so wanted to reference it in context but not critical to this article

 * _[WebDriver](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/webdriver-chromium/c-sharp)_ - a W3C standard wire protocol for out-of-process clients to remotely instruct browsers.
 * _[DevTools Protocol](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/devtools-protocol-chromium)_ - directly inspect, debug, and profile,  Chromium-based browsers like Edge.
 * _[Puppeteer](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/puppeteer)_ - a Node.js _library_ for browser automation using DevTools Protocol.
 * _[Playwright](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/playwright)_ - _framework_ for reliable end-to-end testing of modern web apps, across browsers.
 * _[webhint](https://aka.ms/learn-PWA/30Days-3.6/docs.microsoft.com/en-us/microsoft-edge/test-and-automation/webhint)_ - linter for accessibility, performance, security, PWA and cross-browser compatibility.

We'll focus on [_Playwright_](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev) in this post - but do explore the other resources to learn how they fit into your testing toolkit.

---

## Test Pyramid

> To craft an agile testing strategy, it helps to use visual metaphors like this [practical test pyramid](https://aka.ms/learn-PWA/30Days-3.6/www.pearson.com/us/higher-education/program/Cohn-Succeeding-with-Agile-Software-Development-Using-Scrum/PGM201415.html) that organizes our thinking around two guiding principles: (1) Write tests at various granularities, and (2) Write _fewer_ tests as you go higher up the pyramid.

Reviewer: ```suggestion
To craft an Agile testing strategy, it helps to use visual metaphors like this [practical test pyramid](https://aka.ms/learn-PWA/30Days-3.6/www.pearson.com/us/higher-education/program/Cohn-Succeeding-with-Agile-Software-Development-Using-Scrum/PGM201415.html) that organizes our thinking around two guiding principles: (1) Write tests at various granularities, and (2) Write _fewer_ tests as you go higher up the pyramid.
```

What does testing granularity mean? My version of the pyramid has unit testing at the base, with exploratory testing at the top. Test automation helps with the bottom four layers, while manual testing is key to uncovering "gaps" that automated testing might overlook at the top.

> | Level | Type | Description |

Author: Yes! I had a visual in mind and need to create the image - let me see if I can get that done later today and we can replace it.

|:---|:---|:---|
| 4 | Exploratory Testing | _Manual testing_ (freeform, unstructured, creative) to uncover unforeseen issues, or test edge cases and experimental features. |
| 3 | End-to-End Testing| Tests complete application specification (all stories, all scripts) with realistic test data and measurable outcomes for validation. |
| 2 | UI Testing| Tests user interactions and workflow against a specified story. Validates behavior, layouts, usability, accessibility etc.|

> | 1 | Integration Testing | Tests that multiple components work together as intended, where some are external depenendies for your application. |

Reviewer 2: ```suggestion
| 1 | Integration Testing | Tests that multiple components work together as intended, where some are external dependencies for your application. |
```

| 0 | Unit Testing | Tests that an independent component of work functions as intended.|

> In implementing the strategy, we want _many small, fast tests_ at the base, and _fewer more coarse-grained tests_ as we go higher. There are numerous unit testing libraries and tools that make the first objective achievable. But end-to-end test automation can be ["notoriously flaky"](https://aka.ms/learn-PWA/30Days-3.6/martinfowler.com/articles/practical-test-pyramid.html#TheImportanceOftestAutomation) given the many _browser quirks, timing issues, animations, and unexpected UI interactions_ that we need to deal with.

Author: ```suggestion
In implementing the strategy, we want _many small, fast tests_ at the base, and _fewer more coarse-grained tests_ as we go higher. There are numerous unit testing libraries and tools that make the first objective achievable. But end-to-end test automation can be ["notoriously flaky"](https://aka.ms/learn-PWA/30Days-3.6/martinfowler.com/articles/practical-test-pyramid.html#TheImportanceOftestAutomation) given the many "browser quirks, timing issues, animations, and unexpected UI interactions" that we need to deal with.
```

> In implementing the strategy, we want _many small, fast tests_ at the base, and _fewer more coarse-grained tests_ as we go higher. There are numerous unit testing libraries and tools that make the first objective achievable. But end-to-end test automation can be ["notoriously flaky"](https://aka.ms/learn-PWA/30Days-3.6/martinfowler.com/articles/practical-test-pyramid.html#TheImportanceOftestAutomation) given the many _browser quirks, timing issues, animations, and unexpected UI interactions_ that we need to deal with.

Author: We can use quotes that - it was a literal quote from the referenced article. And yes I need to do less <em> 😶 - will fix

This is where having a reliable testing framework like Playwright really helps. Its [test hierarchies](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/api/class-suite#suite-tests) make it easy to compose high-level _test suites_ from low-level _test cases_ - and use _test configurations_ with multiple _test projects_ to create maximum flexibility. The [test runner](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/intro) can focus on execution, running tests in parallel for efficiency, using multiple projects for cross-browser coverage, and capturing traces for post-execution analysis.

> ---

Author: ```suggestion
```

## Testing PWA

Next, let's think about the _functional_ specification of these tests - what kinds of features (unit testing) and interaction workflows (end-to-end testing) should we consider for Progressive Web Apps? Here's one way to approach this:

 * **PWAs are web apps.** Start with existing ["strategies for carrying out testing"](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Testing_strategies) for web apps, and identify the optimal testing configurations (target devices, browsers) for your intended audience.

>  * **PWAs have additional [desirable characteristics](../core-concepts/01.md)** to validate. We know from the [previous post](05.md) that _auditing tools_ are ideal here and can be integrated into a test automation workflow it they expose a programmable interface. Ex: [playwright-lighthouse](https://github.com/abhinaba-ghosh/playwright-lighthouse) makes it easy to run Lighthouse audits as part of your end-to-end testing workflow.

Reviewer 2: ```suggestion
 * **PWAs have additional [desirable characteristics](../core-concepts/01.md)** to validate. We know from the [previous post](./05.md) that _auditing tools_ are ideal here and can be integrated into a test automation workflow it they expose a programmable interface. Ex: [playwright-lighthouse](https://github.com/abhinaba-ghosh/playwright-lighthouse) makes it easy to run Lighthouse audits as part of your end-to-end testing workflow.
```

 * It has multi-language support - including TypeScript, JavaScript, Java, Python and .NET.
 * No flaky testing - with [auto-waiting](https://playwright.dev/docs/actionability) and [web-first assertions](https://playwright.dev/docs/test-assertions) eliminating messy timeouts.
 * No limits - aligns with modern web architectures, runs out-of-process, and supports multi-everything.
 * Powerful tools - for authoring, debugging, profiling, and reporting tests and execution.
 * Full isolation - with browser contexts ready in ms, fast execution and [parallelization](https://playwright.dev/docs/test-parallel) options.

Want to get hands-on experience with the various capabilities? Check out the [demo.playwright](https://github.com/MarcusFelling/demo.playwright) repository for code examples showcasing test scenarios for accesibility, android, authentication, performance (DevTools & Lighthouse), and visual-comparison. Plus, learn how [Continuous Integration](https://playwright.dev/docs/ci) options like GitHub Actions work seamlessly with Playwright, right out of the box.

---

## Using Playwright

> Plauwright is currently at v1.19, which means you have two options to kickstart your Playwright journey.

Reviewer 2: ```suggestion
Playwright is currently at v1.19, which means you have two options to kickstart your Playwright journey.
```

 * Use the [Getting Started](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/intro) guide to use Playwright from the command line.
 * Use the [Playwright Test for VSCode](https://aka.ms/learn-PWA/30Days-3.6/ms-playwright.playwright) extension (_in preview_) to use Playwright from the IDE.

For now, we'll use the first option since the extension is still in preview and works only with the latest version of Playwright Test. In an earlier post, I [PWA-enabled an existing web app](../core-concepts/07.md). Today, I'll revisit that project and instrument it for end-to-end testing with Playwright. Ready? Let's go.

### 1. Install Playwright

Since I'm scaffolding Playwright tests for an _existing_ project, I'll use the command below. If you are _creating_ a new project, just add the project name as the final argument.

```javascript
$ npm init playwright
```

> The output looks as shown (compressed for clarity). Note that you can choose to get a GitHub Actions workflow instrumented for you during setup. The scaffolding process may take a while the first time since it installs Playwright (Test runner and library) _and_ all supported browsers (Chromium, Firefox and Webkit) by default.

Reviewer: ```suggestion
The output looks as shown (compressed for clarity). Note that you can choose to get a GitHub Actions workflow instrumented for you during setup. The scaffolding process may take a while the first time since it installs Playwright (Test runner and library) and all supported browsers (Chromium, Firefox and Webkit) by default.
```

```
$ npx playwright test
Running 75 tests using 3 workers
...
  75 passed (35s)

To open last HTML report run:
  npx playwright show-report
```

### 3. View the HTML Report

Excellent! The default tests ran and passed! And hey - we have a report so let's check that out!

```javascript
$ npx playwright show-report

  Serving HTML report at http://127.0.0.1:9323. Press Ctrl+C to quit.
```

The report looks something like this. Note how it provides execution time stats on a _per-test (case) + per-browser (project)_ basis, and for the spec (test suite) as a whole.

![Playwright HTML Report](_media/6-playwright-report.png)

 The HTML reporting format allows you to drill down further into test steps, and break down the time taken for executing each test action. This is what the report looks like when you click on the first line item, and drill down into  details. For instance, we can see the test step involved, along with the actual code that was executed from the script.

![Playwright HTML Report](_media/6-playwright-report-detail.png)

### 4. Validate GitHub Actions workflow

> Great! We validated the test script locally and checked that it ran against all browser targets _and_ generated a report that we were able to view locally. The setup step also created a `playwright.yml` file with the GitHub actions workflows for running Playwright tests _and_ uploading reporting artifacts for post-run analysis. 

Reviewer: ```suggestion
Great! We validated the test script locally and checked that it ran against all browser targets _and_ generated a report that we were able to view locally. The setup step also created a `playwright.yml` file with the GitHub actions workflows for running Playwright tests and uploading reporting artifacts for post-run analysis. 
```

 * [playwright.config.js](https://aka.ms/learn-PWA/30Days-3.6/github.com/fearlessly-dev/recipe-me/blob/add-playwright/playwright.config.js) = Test configuration

Merging that branch into main triggers [the GitHub action](https://aka.ms/learn-PWA/30Days-3.6/github.com/fearlessly-dev/recipe-me/actions/runs/1919689804) as anticipated, leading to the entire suite of tests being run in the cloud. The test run took `5m 40s` with the bulk of the time (~4m) used for installing Playwright and browsers. We can also see the [HTML Report Artifact](https://aka.ms/learn-PWA/30Days-3.6/github.com/fearlessly-dev/recipe-me/suites/5498490318/artifacts/175855529) available for download and local viewing.

🎉 Time to celebrate! We have an **end-to-end testing harness setup with continuous integration using GitHub Actions** for our project. It's time to start customizing the test script and configuration to be more aligned to _your_ project needs. We don't have time to dive into details, but let's take a quick look at the two key files to know.

### 5. Inspect the Configuration File

> The Playwright _configuration file_ lets you specify [test options](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/api/class-testoptions) for running multiple test projects concurrently, and [execution options](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/api/class-testconfig) to control how those tests are executed by Playwright. The [playwright.config.js](https://aka.ms/learn-PWA/30Days-3.6/github.com/fearlessly-dev/recipe-me/blob/add-playwright/playwright.config.js) file in the default setup looks something like this (cleaned up for clarity):

Reviewer: ```suggestion
The Playwright configuration file lets you specify [test options](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/api/class-testoptions) for running multiple test projects concurrently, and [execution options](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/api/class-testconfig) to control how those tests are executed by Playwright. The [playwright.config.js](https://aka.ms/learn-PWA/30Days-3.6/github.com/fearlessly-dev/recipe-me/blob/add-playwright/playwright.config.js) file in the default setup looks something like this (cleaned up for clarity):
```

We can now customize this in many ways. Some options to try:
 * Set [trace](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/trace-viewer#recording-a-trace) to `on` to trigger recording, then view the data with Trace Viewer. [Read this post](https://aka.ms/learn-PWA/30Days-3.6/dev.to/azure/004-tool-talk-hello-trace-viewer-2a61) for more details.
 * Add [emulator targets](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/test-configuration#emulation) to `projects` - validate responsiveness of app experiences.
 * Set [screenshot](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/test-configuration#automatic-screenshots) to `on` - make Playwright Test capture screenshots after every test for analysis.
 * Set [video](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/test-configuration#record-video) to `on` - make Playwright Test record videos after every test for analysis.

Note: Every such action has a related cost and complexity. It's good to experiment with various configuration options in your local test environment - then use the reports to iterate for efficiency in time and resource usage.

### 6. Inspect the Test Script

> Start by looking at the basic [first test](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/next/intro#first-test) script to understand key aspects of the test specification. Here is what that looks like:

Author: ```suggestion
Start by looking at the [first test](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/next/intro#first-test) script to understand key aspects of the test specification. Here is what that looks like:
```

});
```

Some observations:
 * `page` is the most common [fixture](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/api/class-fixtures#fixtures-page) used in tests, creating an isolated context for running that test.
 * `page.locator` creates a [Locator](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/api/class-locator), a view into page based on an associated selector.
  * `expect(..).toHaveText(..)` is an example of a web-first [Assertion](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/test-assertions) with convenience async matchers that enforce auto-waiting for reliable testing.

You can learn more about these and other core concepts like test hooks, browser contexts, and explore the Playwright API, by scanning the [documentation](https://aka.ms/learn-PWA/30Days-3.6/playwright.dev/docs/intro#installation). Now, look at the scaffolded [e2e-tests/example.spec.js](https://aka.ms/learn-PWA/30Days-3.6/github.com/fearlessly-dev/recipe-me/blob/add-playwright/e2e-tests/example.spec.js) and see if you can get an understanding of how these concepts translate to more complex end-to-end testing specifications.

### 7. Next Steps 

> The default scaffolded test provides a comprehensive end-to-end testing example for a _ToDoMVC_ application showcasing various features of the Playwright Test API. However, this may be overwhelming for a beginner. Instead, try these steps:

Author: ```suggestion
The default scaffolded test provides a comprehensive end-to-end testing example fora typical "TODO" list  application showcasing various features of the Playwright Test API. However, this may be overwhelming for a beginner. Instead, try these steps:
```
